lib/cfa/augeas_parser/writer.rb

```diff
@@ -0,0 +1,155 @@
+module CFA
+  # Goal of this class is to write back to augeas data stored in {AugeasTree}.
+  # It tries to only required changes, as augeas keeps inside flag if it is
+  # modified and if not modified, then part of file is kept untouched.
+  # @note internal only, unstable API
+  # @private
+  class AugeasWriter
+    # @param aug result of Augeas.create
+    def initialize(aug)
+      @aug = aug
+    end
+
+    # write to augeas data in tree to given prefix
+    # @param prefix [String] where to write tree in augeas
+    # @param tree [CFA::AugeasTree] tree to write
+    def write(prefix, tree)
+      last_valid_entry_path = nil
+      tree.data(filtered: false).each do |entry|
+        key, path = key_and_path(tree, entry, prefix)
+        process_operation(tree, entry, last_valid_entry_path, key, path)
+        last_valid_entry_path = path if entry[:operation] != :remove
+      end
+    end
+
+  private
+
+    attr_reader :aug
+
+    # returns available number for given collection that can be used to write
+    # @param tree [CFA::AugeasTree] tree where collection is defined
+    # @param key [String] collection name
+    def new_array_number(tree, key)
+      all_elements = tree.data(filtered: false)
+                         .select { |e| e[:key] == key }
+      nums = all_elements.map do |entry|
+        entry[:orig_key] ? entry[:orig_key][/^.*\[(\d+)\]$/, 1].to_i : 0
+      end
+      nums.max ? (nums.max + 1) : 1
+    end
+
+    # construct augeas key and whole path where to write for collection
+    def key_and_path_array(tree, entry, prefix)
+      if entry[:orig_key]
+        key = entry[:orig_key]
+      else
+        strip_key = entry[:key]
+        key = strip_key[0..-2] + new_array_number(tree, strip_key).to_s + "]"
```

Comment:
no, as previously it was "key[]" and we remove ending "]" and append "<number>]"